internal/backend/workflows/common.go

```diff
@@ -0,0 +1,141 @@
+package workflows
+
+import (
+	"context"
+	"fmt"
+	"maps"
+
+	"go.autokitteh.dev/autokitteh/internal/backend/temporalclient"
+	"go.autokitteh.dev/autokitteh/internal/kittehs"
+	"go.autokitteh.dev/autokitteh/sdk/sdkservices"
+	"go.autokitteh.dev/autokitteh/sdk/sdktypes"
+	"go.temporal.io/sdk/workflow"
+	"go.uber.org/fx"
+	"go.uber.org/zap"
+)
+
+const (
+	EventsWorkflow        = "events_workflow"
+	SchedulerWorkflow     = "scheduler_workflow"
+	TaskQueueName         = "events-task-queue"
+	ScheduleTaskQueueName = "schedule-task-queue"
+
+	DispatcherWorkerID = "dispatcher-worker"
+	SchedulerWorkerID  = "scheduler-worker"
+)
+
+var (
+	eventInputsSymbolValue   = sdktypes.NewSymbolValue(kittehs.Must1(sdktypes.ParseSymbol("event")))
+	triggerInputsSymbolValue = sdktypes.NewSymbolValue(kittehs.Must1(sdktypes.ParseSymbol("trigger")))
+	dataSymbolValue          = sdktypes.NewSymbolValue(kittehs.Must1(sdktypes.ParseSymbol("data")))
+)
+
+type Services struct {
+	fx.In
+
+	Connections  sdkservices.Connections
+	Deployments  sdkservices.Deployments
+	Events       sdkservices.Events
+	Integrations sdkservices.Integrations
+	Projects     sdkservices.Projects
+	Triggers     sdkservices.Triggers
+	Sessions     sdkservices.Sessions
+	Envs         sdkservices.Envs
+}
+
+type Workflow struct {
+	Z        *zap.Logger
+	Services Services
+	Tmprl    temporalclient.Client
+}
+
+type SessionData struct {
+	Deployment            sdktypes.Deployment
+	CodeLocation          sdktypes.CodeLocation
+	Trigger               sdktypes.Trigger
+	AdditionalTriggerData map[string]sdktypes.Value
+}
+
+// used by both dispatcher and scheduler
+func CreateSessionsForWorkflow(event sdktypes.Event, sessionsData []SessionData) ([]*sdktypes.Session, error) {
+	// DO NOT PASS Memo. It is not intended for automation use, just auditing.
+	eventInputs := event.ToValues()
+	eventStruct, err := sdktypes.NewStructValue(eventInputsSymbolValue, eventInputs)
+	if err != nil {
+		return nil, fmt.Errorf("start session: event: %w", err)
+	}
+
+	inputs := map[string]sdktypes.Value{
+		"event": eventStruct,
+		"data":  eventInputs["data"],
+	}
+
+	sessions := make([]*sdktypes.Session, len(sessionsData))
+	for i, sd := range sessionsData {
+		if t := sd.Trigger; t.IsValid() {
+			inputs = maps.Clone(inputs)
```

Comment:
this is a bug - it'll clone and reclone the previous iteration. you want to clone the original inputs in line 68.

Reply:
not sure I understand. Could u plz elaborate?